rpc/handlers.go

```diff
@@ -1062,3 +1062,67 @@ func (h *Handler) TransactionStatus(hash felt.Felt) (Status, *jsonrpc.Error) {
 
 	return status, nil
 }
+
+func (h *Handler) EstimateFee(broadcastedTxns []BroadcastedTransaction, id BlockID) ([]FeeEstimate, *jsonrpc.Error) {
+	var estimates []FeeEstimate
+	state, closer, err := h.stateByBlockID(&id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+	defer func() {
+		if closeErr := closer(); closeErr != nil {
+			h.log.Errorw("Failed to close state in starknet_estimateFee", "err", closeErr)
+		}
+	}()
+
+	header, err := h.blockHeaderByID(&id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+
+	// todo: remove after the next release
+	if header.GasPrice == nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, "no gas price history for the given block")
+	}
+
+	var txns []core.Transaction
+	var classes []core.Class
+
+	for idx := range broadcastedTxns {
+		txn, declaredClass, aErr := adaptBroadcastedTransaction(&broadcastedTxns[idx], h.network)
+		if aErr != nil {
+			return nil, jsonrpc.Err(jsonrpc.InvalidParams, aErr.Error())
+		}
+
+		txns = append(txns, txn)
+		if declaredClass != nil {
+			classes = append(classes, declaredClass)
+		}
+	}
+
+	blockNumber := header.Number
+	if id.Pending {
+		height, hErr := h.bcReader.Height()
+		if hErr != nil {
+			return nil, ErrBlockNotFound
+		}
+		blockNumber = height + 1
+	}
+
+	gasesConsumed, err := vm.Execute(txns, classes, blockNumber, header.Timestamp, header.SequencerAddress, state, h.network)
+	if err != nil {
+		rpcErr := *ErrContractError
+		rpcErr.Data = err.Error()
+		return nil, &rpcErr
+	}
+
+	for _, gasConsumed := range gasesConsumed {
+		estimates = append(estimates, FeeEstimate{
+			GasConsumed: gasConsumed,
+			GasPrice:    header.GasPrice,
+			OverallFee:  new(felt.Felt).Mul(gasConsumed, header.GasPrice),
```

Comment:
Pending blocks don't have `GasPrice`, so we should set an arbitrary number for pending block

Reply:
Please ignore this, feeder gateway returns the price but rpc doesn't so that's why I was confused.